nullaway/src/main/java/com/uber/nullaway/handlers/fieldcontract/RequiresNonNullHandler.java

```diff
@@ -0,0 +1,203 @@
+            .stream()
+            .map(RequiresNonNullHandler::trimReceiver)
+            .collect(Collectors.toSet());
+    if (overriddenFieldNames.containsAll(overridingFieldNames)) {
+      return;
+    }
+    overridingFieldNames.removeAll(overriddenFieldNames);
+    StringBuilder errorMessage = new StringBuilder();
+    errorMessage.append(
+        "precondition inheritance is violated, method in child class cannot have a stricter precondition than its closest overridden method, adding @requiresNonNull for fields [");
+    Iterator<String> iterator = overriddenFieldNames.iterator();
+    while (iterator.hasNext()) {
+      errorMessage.append(iterator.next());
+      if (iterator.hasNext()) {
+        errorMessage.append(", ");
+      }
+    }
+    errorMessage.append("] makes this method precondition stricter");
+    reportMatch(
+        analysis,
+        state,
+        tree,
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.ANNOTATION_VALUE_INVALID, errorMessage.toString()));
+  }
+
+  /**
+   * It checks whether all class fields given in {@code RequiresNonNull} parameter are @NonNull at
+   * call site.
+   */
+  @Override
+  public void onMatchMethodInvocation(
+      NullAway analysis,
+      MethodInvocationTree tree,
+      VisitorState state,
+      Symbol.MethodSymbol methodSymbol) {
+    Set<String> fieldNames = getFieldNamesFromAnnotation(methodSymbol);
+    if (fieldNames == null) {
+      super.onMatchMethodInvocation(analysis, tree, state, methodSymbol);
+      return;
+    }
+    fieldNames =
+        fieldNames.stream().map(RequiresNonNullHandler::trimReceiver).collect(Collectors.toSet());
+    for (String fieldName : fieldNames) {
+      Symbol.ClassSymbol classSymbol = ASTHelpers.enclosingClass(methodSymbol);
+      assert classSymbol != null
+          : "can not find the enclosing class for method symbol: " + methodSymbol;
+      List<Element> receivers = null; // null receiver means (this) is the receiver.
+      if (tree.getMethodSelect() instanceof MemberSelectTree) {
+        MemberSelectTree memberTree = (MemberSelectTree) tree.getMethodSelect();
+        if (memberTree != null) {
+          receivers = getReceiverTreeElements(memberTree.getExpression());
+        }
+      }
+      Element field = getFieldFromClass(classSymbol, fieldName);
+      assert field != null
+          : "Could not find field: [" + fieldName + "]" + "for class: " + classSymbol;
+      AccessPath accessPath = AccessPath.fromFieldAndBase(receivers, field);
+      Nullness nullness =
+          analysis
+              .getNullnessAnalysis(state)
+              .getNullnessOfAccessPath(
+                  new TreePath(state.getPath(), tree), state.context, accessPath);
+      if (NullabilityUtil.nullnessToBool(nullness)) {
+        reportMatch(
+            analysis,
+            state,
+            tree,
+            new ErrorMessage(
+                ErrorMessage.MessageTypes.ANNOTATION_VALUE_INVALID,
```

Comment:
I think we want a different error message type here than `ANNOTATION_VALUE_INVALID`.  Maybe something like `PRECONDITION_NOT_SATISFIED`?  We could then have `POSTCONDITION_NOT_SATISFIED` for when an `@EnsuresNonNull` annotation is invalid.  Maybe you could see what message types the contract-checking handler is using?